org-code-javabuilder/protocol/src/main/java/org/code/protocol/GlobalProtocol.java

```diff
@@ -37,4 +50,12 @@ public OutputAdapter getOutputAdapter() {
   public InputAdapter getInputAdapter() {
     return this.inputAdapter;
   }
+
+  public String generateAssetUrl(String filename) {
+    return String.format("%s/v3/assets/%s/%s", this.dashboardHostname, this.channelId, filename);
+  }
+
+  public String generateSourcesUrl() {
+    return String.format("%s/v3/sources/%s", this.dashboardHostname, this.channelId);
```

Comment:
My one worry about generating the URLs on Javabuilder is this:
The way we access S3 is in general an area of tech debt that we hope to eventually clean up down the line. I would guess the URLs might change as part of that work. I don't think that should prevent us from adding this code here, but it might be worth adding a note to https://github.com/code-dot-org/code-dot-org/blob/staging/shared/middleware/files_api.rb explaining the dependency. Then future engineers would know to set up a redirect or to ensure all dependencies have been tracked down before changing the URLs.

Reply:
Sounds good. I'll add a comment in my related code-dot-org PR for now. We can revisit if we want to move the url generation into `javabuilder_sessions_controller` instead in the future. I would guess if we change how we access S3 in the future we might have the update javabuilder anyway since it is also accessing S3 (it might not be as simple as replacing a url if we add extra auth for example).